DuckDuckGo/BookmarkSectionDataSources.swift

```diff
@@ -0,0 +1,251 @@
+
+        guard let item = bookmarkItem(at: index) else { return }
+        bookmarksManager.delete(item)
+    }
+
+}
+
+protocol BookmarksSectionDataSourceDelegate: AnyObject {
+    func bookmarksSectionDataSourceDidRequestViewControllerForDeleteAlert(_ bookmarksSectionDataSource: BookmarksSectionDataSource) ->
+    UIViewController
+}
+
+class BookmarksSectionDataSource: BookmarkItemsSectionDataSource {
+    
+    private let bookmarksManager: BookmarksManager
+    
+    private let parentFolder: BookmarkFolder?
+    weak var delegate: BookmarksSectionDataSourceDelegate?
+    
+    private func bookmarkItems() -> [BookmarkItem] {
+        if let folder = parentFolder {
+            return folder.children?.array as? [BookmarkItem] ?? []
+        } else {
+            return bookmarksManager.topLevelBookmarkItems
+        }
+    }
+    
+    var numberOfRows: Int {
+        return max(1, bookmarkItems().count)
+    }
+    
+    var isEmpty: Bool {
+        bookmarkItems().count == 0
+    }
+    
+    init(parentFolder: BookmarkFolder?, delegate: BookmarksSectionDataSourceDelegate?, bookmarksManager: BookmarksManager) {
+        self.parentFolder = parentFolder
+        self.delegate = delegate
+        self.bookmarksManager = bookmarksManager
+    }
+    
+    func bookmarkItem(at index: Int) -> BookmarkItem? {
+        let items = bookmarkItems()
+        if items.count <= index {
+            return nil
+        }
+        return items[index]
+    }
+    
+    func cell(_ tableView: UITableView, forIndex index: Int) -> UITableViewCell {
+        if isEmpty {
+            tableView.separatorColor = parentFolder != nil  ? .clear : UIColor(named: "BookmarksCellSeperatorColor")
+            return createEmptyCell(tableView, forIndex: index)
+        } else {
+            tableView.separatorColor = UIColor(named: "BookmarksCellSeperatorColor")
+            return createCell(tableView, withItem: bookmarkItem(at: index))
+        }
+    }
+    
+    func navigationTitle() -> String? {
+        return parentFolder?.title
+    }
+
+    func title() -> String? {
+        return parentFolder == nil ? UserText.sectionTitleBookmarks : nil
+    }
+    
+    func commit(_ tableView: UITableView, editingStyle: UITableViewCell.EditingStyle, forRowAt index: Int, section: Int) {
+        guard editingStyle == .delete else { return }
+
+        guard let item = bookmarkItem(at: index) else { return }
+        if let delegate = delegate,
+            let folder = item as? BookmarkFolder,
+            (folder.children?.count ?? 0) > 0 {
+            let title = String(format: UserText.deleteBookmarkFolderAlertTitle, folder.title ?? "")
+            let count = folder.children?.count ?? 0
+            let messageString: String
+            if count == 1 {
+                messageString = UserText.deleteBookmarkFolderAlertMessageSingular
+            } else {
+                messageString = UserText.deleteBookmarkFolderAlertMessagePlural
```

Comment:
Shouldn't this use strings dict to properly translate plural form? See `TabSwitcherViewController:refreshTitle()` for reference.

Also, if I have:
Bookmarks -> Folder A (one subfolder) -> Folder B (10 subfolders) message for removal of 'Folder A' will only warn me about removing one item, not a bunch of these from a subtree.